### PR TITLE
feat(solc): create compilation unit for files that share solc version/ settings

### PR DESCRIPTION
**Pull request by 0xalpharush**
## Motivation

Currently, the cache file stores the compiler version and its settings individually for each file, and it can be ambiguous what compilation artifacts are a "unit"; that is, they were determined to be compatible and/or necessary e.g. imports and then compiled with the same solc version and compiler settings. This is mainly a concern when multiple compiler versions are used and the artifacts can then contain duplicate id's in the source mappings (see https://github.com/crytic/slither/issues/1213). 

Ref https://github.com/foundry-rs/foundry/issues/3450
## Solution
This PR changes the data model of the artifacts to instead store a unique identifier representing that the source files were compiled together (`CompilationUnitId`). This identifier maps back to a `CompilationUnit` which stores the version, configuration, and source files' paths. Rather than redundantly store the `SolcConfig` for each `CacheEntry`, it is now looked up by its `CompilationUnitId`.  Now, the each compilation unit can be parsed individually and lookup where the artifact is in the cache entry's `files`. This [diff](https://github.com/0xalpharush/foundry/commit/61f054038329e7a2eae175993487bdf4c3ab6ef2) in Foundry is pretty small fwiw.

I'm new to Rust and also may have overlooked some design decision that my changes are incompatible with so I'm very open to feedback!

## PR Checklist

-   [x] Added Tests
-   [X] Added Documentation
-   [ ] Updated the changelog
-   [X] Breaking changes

### CODE REVIEW

**Comment by plotchy** [comment]
The attached slither issue shows one of the most ubiquitous, difficult to diagnose bugs in slither. very nice job pursuing a fix

File: ethers-solc/src/artifacts/mod.rs
Comment (mattsse) [inline ethers-solc/src/artifacts/mod.rs:516]: what's the reason for the manual impls here?

manual hash impl for settings and including the metadatahash directly makes sense tho

File: ethers-solc/src/cache.rs
Comment (mattsse) [inline ethers-solc/src/cache.rs:35]: should this take the hasher outcome u64 instead of String and print hex via Display impl?

Comment (mattsse) [inline ethers-solc/src/cache.rs:35]: needs some docs

File: ethers-solc/src/cache.rs
Comment (mattsse) [inline ethers-solc/src/cache.rs:54]: let's make this ordered?

```suggestion
    pub source_units: BTreeMap<PathBuf>,
```

File: ethers-solc/src/cache.rs
Comment (mattsse) [inline ethers-solc/src/cache.rs:66]:
```suggestion
    pub compilation_units: BTreeMap<CompilationUnitId, CompilationUnit>,
```

File: ethers-solc/src/config.rs
Comment (mattsse) [inline ethers-solc/src/config.rs:759]: can just youse derived impls instead?

File: ethers-solc/src/cache.rs
Comment (mattsse) [inline ethers-solc/src/cache.rs:39]: can take args as a reference?

**Review by mattsse — CHANGES_REQUESTED** [review]
thanks for this.

Given that this will improve slither integration, I'm supportive of this @gakonst.

on thing I'm not certain of is, how this should/will affect partial recompilation on changes. 
Currently, we only include the necessary files needed by a modified file. would the compileIdunit still be valid if the settings did not change?

compiling the entire unit on a single change would not be feasible imo

**Comment by gakonst** [comment]
> Given that this will improve slither integration, I'm supportive of this @gakonst.

Same. Great change, thanks for taking this on @0xalpharush.

> on thing I'm not certain of is, how this should/will affect partial recompilation on changes.
> Currently, we only include the necessary files needed by a modified file. would the compileIdunit still be valid if the settings did not change?
> compiling the entire unit on a single change would not be feasible imo

Agreed. Let's add a test for this to check what happens to partial recompilation?

**Review by gakonst — CHANGES_REQUESTED** [review]
Would like an extra test for the partial recompilation case that Matt mentioned. Appreciate you tackling this.

File: ethers-solc/src/cache.rs
Comment (0xalpharush) [reply ethers-solc/src/cache.rs]: https://github.com/gakonst/ethers-rs/pull/1964/commits/80c3e3e4959bcb975d43af313b292d6cd2c68d1e

File: ethers-solc/src/artifacts/mod.rs
Comment (0xalpharush) [reply ethers-solc/src/artifacts/mod.rs]: I derived Hash impls where necessary https://github.com/gakonst/ethers-rs/pull/1964/commits/94e9415b87d21e396ca4a1d1bce38dd3519e3237
> manual hash impl for settings and including the metadatahash directly makes sense tho
Are you referring to `SettingsMetadata`?

File: ethers-solc/src/cache.rs
Comment (0xalpharush) [reply ethers-solc/src/cache.rs]: https://github.com/gakonst/ethers-rs/pull/1964/commits/28bf56149c98ed8a87319f0ba2869d3481225159

File: ethers-solc/src/cache.rs
Comment (0xalpharush) [reply ethers-solc/src/cache.rs]: https://github.com/gakonst/ethers-rs/pull/1964/commits/4439d1435cf44b8cb76d7d82ebc52a051ee99e47

**Comment by 0xalpharush** [comment]
> Currently, we only include the necessary files needed by a modified file. would the compileIdunit still be valid if the settings did not change?

@mattsse AFAICT I didn't modify the detection of "clean" and "dirty" sources (there is an existing test for [this](https://github.com/gakonst/ethers-rs/blob/799f752e55e80aa8d6357321d4f5f7ebf66daa8a/ethers-solc/src/compile/project.rs#L726)). The only time an entire compilation unit should be recompiled is if the settings are changed for the entire unit e.g. an incompatible solidity version, creating another compilation unit derived from the incompatible settings. I added a [test case](https://github.com/gakonst/ethers-rs/pull/1964/commits/98d823c74a26c386d276d87dd0e74aae617de8dc) to show that the compilation unit id remains valid if only a file's contents are changed and that only the dirty source is recompiled.

**Comment by gakonst** [comment]
@mattsse PTAL I think this is starting to look good

File: ethers-solc/src/cache.rs
Comment (mattsse) [reply ethers-solc/src/cache.rs:60]: this is written to disk, so we want consistent order

**Review by mattsse — APPROVED** [review]
thanks for clarifying @0xalpharush 

changes and test look good, let's rebase this PR then the foundry companion, good to merge when the companion passes.

**Comment by 0xalpharush** [comment]
Rebased. I think the ethereum test failure is unrelated:
```
---- nonce_manager stdout ----
thread 'nonce_manager' panicked at 'called `Result::unwrap()` on an `Err` value: MiddlewareError(MiddlewareError(JsonRpcClientError(JsonRpcError(JsonRpcError { code: -32000, message: "replacement transaction underpriced", data: None }))))', ethers-middleware/tests/nonce_manager.rs:42:14
```
Updated the Foundry PR https://github.com/foundry-rs/foundry/pull/3939

**Review by gakonst — APPROVED** [review]
LGTM pending @mattsse approval

File: ethers-solc/src/cache.rs
Comment (0xalpharush) [inline ethers-solc/src/cache.rs:452]: @gakonst One thing I realized chasing down #1995 is that each cache entry should have multiple compilation units instead of just one so maybe the compilation unit should be stored at the artifact level. This would allow the cache to handle having the same version being compiled with different compiler settings e.g. someone doesn't want to use via-IR while testing. I explore this in a separate branch but not sure it's a great solution (https://github.com/0xalpharush/ethers-rs/commit/c378b495f712aae2cc077d6d3e50ab4def1e67bd).

Comment (mattsse) [reply ethers-solc/src/cache.rs:452]: this is a reasonable consideration and would give us a bit more flexibility I think.

can't think of any downsides to moving this to the artifact level

Comment (gakonst) [reply ethers-solc/src/cache.rs:452]: SGTM.

**Comment by gakonst** [comment]
@0xalpharush friendly ping here, no pressure again just dont want this to get stale

**Comment by 0xalpharush** [comment]
Will get back to this when I have the time. I started over holiday break and have been busy recently... If anyone finds a solution to https://github.com/gakonst/ethers-rs/issues/1995, that would be great to have for the time being.

**Comment by gakonst** [comment]
ACK. -thx! #1995 up for grabs.

**Comment by gakonst** [comment]
Closing as stale.

**Pull request closed without merging**
